### PR TITLE
Breaking: Switch to Espree 3.0.0 (fixes #4334)

### LINKED ISSUE #4334: Switch to Espree 3.0.0
When it's ready, we'll switch to Espree 3.0.0, which is the first version powered by Acorn.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27978859-switch-to-espree-3-0-0?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by nzakas**
These are all the changes necessary for ESLint to pass all tests using Espree 3.0.0.

### CODE REVIEW

File: lib/rule-context.js [hunk 1]
```diff
@@ -98,7 +99,7 @@ function RuleContext(ruleId, eslint, severity, options, settings, ecmaFeatures)
      * The read-only ecmaFeatures shared across all rules
      */
     Object.defineProperty(this, "ecmaFeatures", {
-        value: Object.create(ecmaFeatures)
+        value: assign({}, ecmaFeatures)
```
Comment (nzakas) [inline lib/rule-context.js]: Prototype properties were causing problems based on how we pass around `ecmaFeatures`

File: lib/rules/key-spacing.js [hunk 2]
```diff
@@ -142,7 +157,7 @@ module.exports = function(context) {
     function report(property, side, whitespace, expected) {
         var diff = whitespace.length - expected,
             key = property.key,
-            firstTokenAfterColon = context.getTokenAfter(key, 1),
+            firstTokenAfterColon = context.getTokenAfter(getNextColon(key)),
```
Comment (nzakas) [inline lib/rules/key-spacing.js]: Some small alignment differences.

File: tests/lib/eslint.js [hunk 3]
```diff
@@ -2367,7 +2367,7 @@ describe("eslint", function() {
             }, filename);
 
             assert.equal(messages.length, 1);
-            assert.equal(messages[0].message, "Parsing error: Illegal return statement");
+            assert.equal(messages[0].message, "Parsing error: 'return' outside of function");
```
Comment (nzakas) [inline tests/lib/eslint.js]: Different error message format from the parser.

File: tests/lib/rules/comma-spacing.js [hunk 4]
```diff
@@ -61,7 +61,7 @@ ruleTester.run("comma-spacing", rule, {
         {code: "`${[1, 2]}`", ecmaFeatures: {templateStrings: true}},
         "foo(/,/, 'a')",
         "var x = ',,,,,';",
-        "var code = 'var foo = 1, bar = 3;',",
+        "var code = 'var foo = 1, bar = 3;'",
```
Comment (nzakas) [inline tests/lib/rules/comma-spacing.js]: This is actually illegal syntax.

File: tests/lib/rules/key-spacing.js [hunk 5]
```diff
@@ -326,7 +326,7 @@ ruleTester.run("key-spacing", rule, {
         code: "var obj = { [ (a + b) ]:value };",
         options: [{}],
         ecmaFeatures: { objectLiteralComputedProperties: true },
-        errors: [{ message: "Missing space before value for computed key \"(a + b)\".", type: "Identifier", line: 1, column: 24 }]
+        errors: [{ message: "Missing space before value for computed key \"a + b\".", type: "Identifier", line: 1, column: 25 }]
```
Comment (nzakas) [inline tests/lib/rules/key-spacing.js]: Slight difference in location calculation.

File: tests/lib/rules/no-native-reassign.js [hunk 6]
```diff
@@ -29,14 +29,6 @@ ruleTester.run("no-native-reassign", rule, {
     invalid: [
         { code: "String = 'hello world';", errors: [{ message: "String is a read-only native object.", type: "Identifier"}] },
         {
-            code: "({Object = 0, String = 0}) = {};",
```
Comment (nzakas) [inline tests/lib/rules/no-native-reassign.js]: Illegal syntax.

File: tests/lib/rules/no-unused-vars.js [hunk 7]
```diff
@@ -87,7 +87,6 @@ ruleTester.run("no-unused-vars", rule, {
         {code: "const x = 1; const {z: [y = x]} = {}; foo(y);", ecmaFeatures: {blockBindings: true, destructuring: true}},
         {code: "const x = []; const {z: [y] = x} = {}; foo(y);", ecmaFeatures: {blockBindings: true, destructuring: true}},
         {code: "const x = 1; let y; [y = x] = []; foo(y);", ecmaFeatures: {blockBindings: true, destructuring: true}},
-        {code: "const x = 1; let y; ({y = x}) = {}; foo(y);", ecmaFeatures: {blockBindings: true, destructuring: true}},
```
Comment (nzakas) [inline tests/lib/rules/no-unused-vars.js]: Illegal syntax on both changes in this file.

File: tests/lib/rules/sort-vars.js [hunk 8]
```diff
@@ -35,8 +35,8 @@ ruleTester.run("sort-vars", rule, {
         { code: "var A, a;", options: ignoreCaseArgs },
         { code: "var a, B, c;", options: ignoreCaseArgs },
         { code: "var A, b, C;", options: ignoreCaseArgs },
-        { code: "var {a, b, c};", options: ignoreCaseArgs, ecmaFeatures: { destructuring: true }},
-        { code: "var {A, b, C};", options: ignoreCaseArgs, ecmaFeatures: { destructuring: true }},
+        { code: "var {a, b, c} = x;", options: ignoreCaseArgs, ecmaFeatures: { destructuring: true }},
```
Comment (nzakas) [inline tests/lib/rules/sort-vars.js]: Fixing illegal syntax in this file.

File: tests/lib/util/comment-event-generator.js [hunk 9]
```diff
@@ -32,7 +32,7 @@ describe("NodeEventGenerator", function() {
     it("should generate comment events without duplicate.", function() {
         var emitter = new EventEmitter();
         var generator = new NodeEventGenerator(emitter);
-        var code = "//foo\nvar zzz /*aaa*/ = 777\n//bar";
+        var code = "//foo\nvar zzz /*aaa*/ = 777;\n//bar";
```
Comment (nzakas) [inline tests/lib/util/comment-event-generator.js]: This code example was broken due to ASI differences in Espree 2.x. In 3.x, there are no differences so updated just to make it clear.

File: tests/lib/rules/no-native-reassign.js (hunk 6 above)
Comment (mysticatea) [inline tests/lib/rules/no-native-reassign.js]: Does `({Object = 0, String = 0} = {});` work?

Comment (michaelficarra) [inline tests/lib/rules/no-native-reassign.js]: Agreed. Instead of deleting these lines, they can be replaced.

Comment (nzakas) [inline tests/lib/rules/no-native-reassign.js]: I'll add that back.

**Comment by nzakas** [comment]
Merging so we start testing with Espree 3 ASAP.

**Pull request merged**